autoretrieve/autoretrieve.go

```diff
@@ -0,0 +1,322 @@
+package autoretrieve
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/application-research/estuary/util"
+	provider "github.com/filecoin-project/index-provider"
+	"github.com/filecoin-project/index-provider/metadata"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/multiformats/go-multiaddr"
+	"github.com/multiformats/go-multihash"
+	"gorm.io/gorm"
+)
+
+type Autoretrieve struct {
+	gorm.Model
+
+	Handle            string `gorm:"unique"`
+	Token             string `gorm:"unique"`
+	LastConnection    time.Time
+	LastAdvertisement time.Time
+	PrivateKey        string `gorm:"unique"`
+	Addresses         string
+}
+
+type HeartbeatAutoretrieveResponse struct {
+	Handle            string         `json:"handle"`
+	LastConnection    time.Time      `json:"lastConnection"`
+	LastAdvertisement time.Time      `json:"lastAdvertisement"`
+	AddrInfo          *peer.AddrInfo `json:"addrInfo"`
+}
+
+type AutoretrieveListResponse struct {
+	Handle            string         `json:"handle"`
+	LastConnection    time.Time      `json:"lastConnection"`
+	LastAdvertisement time.Time      `json:"lastAdvertisement"`
+	AddrInfo          *peer.AddrInfo `json:"addrInfo"`
+}
+
+type AutoretrieveInitResponse struct {
+	Handle         string         `json:"handle"`
+	Token          string         `json:"token"`
+	LastConnection time.Time      `json:"lastConnection"`
+	AddrInfo       *peer.AddrInfo `json:"addrInfo"`
+}
+
+type SimpleEstuaryMhIterator struct {
+	offset int
+	Mh     []multihash.Multihash
+}
+
+func (m *SimpleEstuaryMhIterator) Next() (multihash.Multihash, error) {
+	if m.offset < len(m.Mh) {
+		hash := m.Mh[m.offset]
+		m.offset++
+		return hash, nil
+	}
+	return nil, io.EOF
+}
+
+// newIndexProvider creates a new index-provider engine to send announcements to storetheindex
+// this needs to keep running continuously because storetheindex
+// will come to fetch advertisements "when it feels like it"
+func NewAutoretrieveEngine(stopCh chan struct{}, tickInterval time.Duration, db *gorm.DB) (*AutoretrieveEngine, error) {
```

Comment:
instead of storing `stopCh` here, the `.Run()` function should probably just take a context like `Run(ctx context.Context)` and listen for `<-ctx.Done()`. a lot nicer if you're also working with other functions that take a context, because they can all be canceled in a consistent way